azure-proto-core/Resources/SystemAssignedIdentity.cs

```diff
@@ -0,0 +1,144 @@
+﻿namespace azure_proto_core.Resources
+{
+    using System;
+    using System.Text.Json;
+    using Azure.Core;
+
+    public class SystemAssignedIdentity
+    {
+        public Guid? TenantId { get; set; }
+
+        public Guid? PrincipalId { get; set; }
+
+        public SystemAssignedIdentity() { }
+
+        public SystemAssignedIdentity(Guid tenantId, Guid principalId)
+        {
+            TenantId = tenantId;
+            PrincipalId = principalId;
+        }
+
+        public int CompareTo(SystemAssignedIdentity other)
+        {
+            if (other == null)
+                return 1;
+
+            if (TenantId.HasValue == false &&
+                PrincipalId.HasValue == false &&
+                other.TenantId.HasValue == false &&
+                other.PrincipalId.HasValue == false)
+            {
+                return 0;
+            }
+
+            if (!TenantId.HasValue && !PrincipalId.HasValue)
+                return -1;
+
+            if (!other.TenantId.HasValue && !other.PrincipalId.HasValue)
+                return 1;
+
+            if (TenantId.Value.CompareTo(other.TenantId.Value) == 1 && PrincipalId.Value.CompareTo(other.PrincipalId.Value) == 1)
+                return 1;
+
+            if (TenantId.Value.CompareTo(other.TenantId.Value) == 0 && PrincipalId.Value.CompareTo(other.PrincipalId.Value) == 0)
+                return 0;
+
+            return -1;
+        }
+
+        public bool Equals(SystemAssignedIdentity other)
+        {
+            if (other == null)
+                return false;
+
+            if (TenantId.HasValue == false &&
+                PrincipalId.HasValue == false &&
+                other.TenantId.HasValue == false &&
+                other.PrincipalId.HasValue == false)
+            {
+                return true;
+            }
+
+            if (!TenantId.HasValue && !PrincipalId.HasValue)
+                return false;
```

Comment:
What if TenatId has value but PrincipleID does not? 

Equals on line 68 would fail no?

Reply:
Then no need to check both right?